osu.Game/Rulesets/Mods/ModChallenge.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using osu.Framework.Bindables;
+using osu.Game.Configuration;
+using osu.Game.Rulesets.Objects;
+using System.Collections.Generic;
+using osu.Game.Beatmaps;
+using osu.Game.Beatmaps.Timing;
+using osu.Game.Scoring;
+using osu.Game.Rulesets.Judgements;
+using osu.Game.Rulesets.Scoring;
+using System.ComponentModel;
+
+namespace osu.Game.Rulesets.Mods
+{
+    public abstract class ModChallenge : ModFailCondition, IApplicableToBeatmap
+    {
+        public override ModType Type => ModType.DifficultyIncrease;
+        public override double ScoreMultiplier => 1;
+        public override bool RequiresConfiguration => true;
+        public override Type[] IncompatibleMods => base.IncompatibleMods.Append(typeof(ModPerfect)).Append(typeof(ModEasyWithExtraLives)).ToArray();
+
+        [SettingSource("Checking interval", "When conditions will be checked.")]
+        public Bindable<ChallengeCheckInterval> CheckingInterval { get; } = new Bindable<ChallengeCheckInterval>();
+
+        protected readonly List<HitObject> HitObjectsBeforeBreaks = new List<HitObject>();
+        protected HitObject LastHitObjectInBeatmap;
+
+        public void ApplyToBeatmap(IBeatmap beatmap)
+        {
+            switch (CheckingInterval.Value)
+            {
+                case ChallengeCheckInterval.AtEnd:
+                    LastHitObjectInBeatmap = beatmap.HitObjects.LastOrDefault();
+                    break;
+
+                case ChallengeCheckInterval.AtBreak:
+                    for (int i = 0; i < beatmap.HitObjects.Count - 1; i++)
+                    {
+                        double inBetweenTime = (beatmap.HitObjects.ElementAtOrDefault(i).GetEndTime() + beatmap.HitObjects.ElementAtOrDefault(i + 1).StartTime) / 2;
```

Comment:
This implicitly assumes that an object is before a break if the midpoint between its end and the start of the next one is inside a break, which I don't think is necessarily correct. Either way I'd like to see this entire method to be deleted entirely for now, so I'm not going to go into the nitty gritty of how to do this properly.